core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -99,21 +99,22 @@ class LogSegment private[log] (val log: FileRecords,
   // volatile for LogCleaner to see the update
   @volatile private var rollingBasedTimestamp: Option[Long] = None
 
+  /* The maximum timestamp and offset we see so far */
+  @volatile private var _maxTimestampAndOffsetSoFar: (Option[Long], Option[Long]) = (None, None)
+  def maxTimestampAndOffsetSoFar_= (timestampAndOffset: (Long, Long)) : Unit = _maxTimestampAndOffsetSoFar = (Some(timestampAndOffset._1), Some(timestampAndOffset._2))
+  def maxTimestampAndOffsetSoFar: (Long,Long) = {
+    if (_maxTimestampAndOffsetSoFar._1.isEmpty || _maxTimestampAndOffsetSoFar._2.isEmpty)
+      _maxTimestampAndOffsetSoFar = (Some(timeIndex.lastEntry.timestamp), Some(timeIndex.lastEntry.offset))
+    (_maxTimestampAndOffsetSoFar._1.get, _maxTimestampAndOffsetSoFar._2.get)
+  }
```

Comment:
Have you considered using a case class instead of `(Option[Long], Option[Long])` here? We could do something like this:

```
case class TimestampAndOffset(timestamp: Long, offset: Long)

object TimestampAndOffset {
  val empty = TimestampAndOffset(RecordBatch.NO_TIMESTAMP, 0L) // Not sure about 0 by default here.
}
```

Then, we could define:

```
@volatile private var _maxTimestampAndOffsetSoFar:  TimestampAndOffset = TimestampAndOffset.empty.

def maxTimestampAndOffsetSoFar: TimestampAndOffset = {
  if (_maxTimestampAndOffsetSoFar == TimestampAndOffset.empty)
    _maxTimestampAndOffsetSoFar = TimestampAndOffset(...)
  _maxTimestampAndOffsetSoFar
}
...
```

One point which is not clear to me is wether we really could use `TimestampAndOffset.empty` as a sentinel or we must wrap it in an `Option`. I need to dig a bit in the code to better understand it here.

Perhaps, `TimestampAndOffset` already exists in the codebase. I haven't checked.

Reply:
Thanks for the review, I switched to case class for clarity, TimestampAndOffset does already exist but is not suitable for this purpose so I named it MaxTimestampAndOffset. It looks like the options are only checked during close() and so MaxOffsetAndTimestamp.empty is valid without them. I have changed the timestamp to -1L to make sure it is not a valid value when unset.

Reply:
There is a `TimestampOffset` too. Can we use that?

Reply:
Indeed, we could use `TimestampOffset` and `TimestampOffset.Unknown`.

Reply:
Nice find! I've updated the PR to use this.